pkg/summarizer/flakiness.go

```diff
@@ -0,0 +1,286 @@
+package summarizer
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/testgrid/internal/result"
+	"github.com/GoogleCloudPlatform/testgrid/pb/state"
+	"github.com/sirupsen/logrus"
+	"vbom.ml/util/sortorder"
+)
+
+var INFRA_REGEX = regexp.MustCompile(`^\w+$`)
+var JAILED_REGEX = regexp.MustCompile(`(JAILED )?(\/?\/?[A-Za-z0-9\/_\-.:}{]+) - \[([A-Za-z0-9\/:\-_.\s]+)\]\s*`)
+
+// Keep in mind that flakiness is measured as out of 100, i.e. 23 not .23
+var DEFAULT_FLAKINESS = 50.0
+var MIN_RUNS = 0
+var HEALTHY_RANGE = map[int]string{
+	20: "red",
+	10: "purple",
+	3:  "orange",
+	0:  "green",
+}
+
+type Result struct {
+	name             string
+	passed           int
+	failed           int
+	flakyCount       int
+	averageFlakiness float64
+	failedInfraCount int
+	infraFailures    map[string]int
+}
+
+// Temporary structs while I decide what will go into summary.proto
+type Healthiness struct {
+	startDate        int
+	endDate          int
+	tests            []TestInfo
+	totalTests       int
+	totalJailedTests int
+	averageFlakiness float64
+	flakyBuckets     []FlakyBucket
+	infraIssues      map[string]int
+	totalConfigs     int
+}
+
+type TestInfo struct {
+	name               string
+	env                string
+	totalRuns          int
+	totalRunsWithInfra int
+	passedRuns         int
+	failedRuns         int
+	failedInfraRuns    int
+	flakyRuns          int
+	flakiness          float64
+	infraInfo          string
+}
+
+type FlakyBucket struct {
+	threshold float64
+	tests     int
+}
+
+// For sorting, primarily intended for map[string]int as implemented below
+type IntString struct {
+	i int
+	s string
+}
+
+func CalculateHealthiness(grid *state.Grid, startTime int, endTime int, tab string) Healthiness {
+	results := parseGrid(grid, startTime, endTime)
+	return analyzeFlakinessFromResults(results, startTime, endTime, tab)
+}
+
+func parseGrid(grid *state.Grid, startTime int, endTime int) []Result {
+	// Get the relevant data for flakiness from each Grid (which represents
+	// a dashboard tab) as a list of Result structs
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	results := make([]Result, 0)
```

Comment:
The pattern I have seen is to use 
```
// results returns a per-column result output channel for each row.
func results(ctx context.Context, rows []*statepb.Row) map[string]<-chan statepb.Row_Result {
	return result.Map(ctx, rows)
}
```
Something like this and iterate over the results channel. Tbh not super well versed in Golang so not sure what the real value is here

Reply:
Done. I changed it to be similar to [gridMetrics](https://github.com/GoogleCloudPlatform/testgrid/blob/master/pkg/summarizer/summary.go#L496) 
where each channel has a chance to generate the next RowResult after the current one is consumed. I think channels work like python generator functions but with concurrency.